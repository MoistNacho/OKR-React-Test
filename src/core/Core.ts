import { observable } from "mobx";

import Dialog from "./Dialog";
import PopUp from "./PopUp";
import Router from "./Router";

export default class Core {
  public dialog: Dialog;

  public router: Router;

  public popUp: PopUp;

  @observable
  public locale = "ko";

  private stores: {
    [key: string]: any;
  } = {};

  constructor() {
    this.dialog = new Dialog();
    this.router = new Router();
    this.popUp = new PopUp();

    // expose core during tests
    if ((window as any).Cypress) {
      (window as any).core = this;
    }
  }

  public addStore = (key: string, store: any) => {
    this.stores[key] = store;
  };

  public getStore = (key: string) => this.stores[key];

  public clearStores = () => {
    this.stores = {};
  };

  public reset = () => this.clearStores();
}
