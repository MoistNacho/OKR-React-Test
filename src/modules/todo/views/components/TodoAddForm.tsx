import { ButtonV2, TextInputV2 } from "@meshkorea/vroong-design-system-web";
import React, { useCallback, useState } from "react";
import styled from "styled-components";

interface TodoAddFormProps {
  onCreate: (value: string) => void;
}

const TodoAddForm = ({ onCreate }: TodoAddFormProps) => {
  const [value, setValue] = useState<string>();

  const handleInputValue = useCallback((v: string) => {
    setValue(v);
  }, []);

  const onSubmit = useCallback(() => {
    onCreate(value!);
    setValue("");
  }, [onCreate, value]);

  return (
    <AddTodoWrap>
      <TextInputV2
        width="400px"
        placeholder="값을 입력해주세요"
        value={value}
        onChange={handleInputValue}
      />
      <ButtonV2 onClick={onSubmit} disabled={!value}>
        확인
      </ButtonV2>
    </AddTodoWrap>
  );
};

export default TodoAddForm;

const AddTodoWrap = styled.div`
  margin: 0 auto;
`;
