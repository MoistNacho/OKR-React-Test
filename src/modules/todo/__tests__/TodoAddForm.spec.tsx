import { BaseTheme } from "@meshkorea/vroong-design-system-web";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import React from "react";
import renderer from "react-test-renderer";
import { ThemeProvider } from "styled-components";

import { CoreProvider } from "core";

import TodoAddForm from "../views/components/TodoAddForm";

describe("<TodoAddForm />", () => {
  it("renders with snapshot", () => {
    const component = renderer.create(
      <CoreProvider>
        <ThemeProvider theme={BaseTheme}>
          <TodoAddForm onCreate={jest.fn()} />
        </ThemeProvider>
      </CoreProvider>,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  describe("Form Test", () => {
    const onCreate = jest.fn();
    let input: HTMLInputElement;
    let submitButton: HTMLButtonElement;

    beforeEach(() => {
      render(
        <CoreProvider>
          <ThemeProvider theme={BaseTheme}>
            <TodoAddForm onCreate={onCreate} />
          </ThemeProvider>
        </CoreProvider>,
      );
      input = screen.getByPlaceholderText("값을 입력해주세요");
      submitButton = screen.getByRole("button", { name: "확인" });
    });

    it("input에 값을 입력시 button의 disabled 값이 false로 변경된다", async () => {
      await userEvent.type(input, "new Item");

      expect(submitButton.disabled).toBe(false);
    });

    it("submit버튼을 클릭시 onCreate함수에 input의 value값이 함께 전달된다", async () => {
      await userEvent.type(input, "new Item");
      await userEvent.click(submitButton);

      expect(onCreate).toHaveBeenCalledWith("new Item");
    });
  });
});
