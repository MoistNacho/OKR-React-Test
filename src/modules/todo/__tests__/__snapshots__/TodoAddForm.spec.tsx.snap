// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoAddForm /> renders with snapshot 1`] = `
<div
  className="sc-bPaAec hDWzBG"
>
  <div
    className="sc-dZTDFv bJPCES"
    width="400px"
  >
    <div
      className="sc-hbWBzy dJsoGE"
      disabled={false}
    >
      <input
        className="sc-nTrUm fNPogZ"
        onChange={[Function]}
        onKeyDown={[Function]}
        onKeyPress={[Function]}
        placeholder="값을 입력해주세요"
        type="text"
        value=""
      />
    </div>
  </div>
  <button
    className="sc-bcXHqe dCAPbY"
    disabled={true}
    onClick={[Function]}
    onKeyPress={[Function]}
    type="button"
  >
    확인
  </button>
</div>
`;
